### PR TITLE
Fixing APIMANAGER-3155 :issues with api grouping based on tags

### PR DESCRIPTION
**Pull request by malinthaprasan**

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 1]
```diff
@@ -1081,14 +1082,45 @@ public int compare(Tag o1, Tag o2) {
 		                                APIConstants.TAGS_INFO_ROOT_LOCATION +
 		                                        "/%s/description.txt";
 		String thumbnailPathPattern = APIConstants.TAGS_INFO_ROOT_LOCATION + "/%s/thumbnail.png";
+
+		//if the tenantDomain is not specified super tenant domain is used
+		if(tenantDomain == null || tenantDomain.trim() == "" ){
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: use StringUtils.isEmpty

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 2]
```diff
@@ -1081,14 +1082,45 @@ public int compare(Tag o1, Tag o2) {
...
+			try {
+				tenantDomain = ServiceReferenceHolder.getInstance().getRealmService().getTenantManager().getSuperTenantDomain();
+			} catch (org.wso2.carbon.user.core.UserStoreException e) {
+				handleException("Cannot get super tenant domain name",e);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: slight formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 3]
```diff
@@ -1081,14 +1082,45 @@ public int compare(Tag o1, Tag o2) {
...
+			ServiceReferenceHolder.getInstance().getRealmService().getTenantManager().getTenantId(tenantDomain);
+			int tenantId = ServiceReferenceHolder.getInstance().getRealmService().getTenantManager()
+			                                     .getTenantId(tenantDomain);
+			RegistryService registryService =
+					ServiceReferenceHolder.getInstance()
+					                      .getRegistryService();
+			govRegistry = registryService.getGovernanceSystemRegistry(tenantId);
+
+		} catch (UserStoreException e) {
+			handleException("Cannot get tenant id for tenant domain name:"+tenantDomain,e);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: slight formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 4]
```diff
@@ -1081,14 +1082,45 @@ public int compare(Tag o1, Tag o2) {
...
+		} catch (RegistryException e) {
+			handleException("Cannot get registry for tenant domain name:"+tenantDomain,e);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: slight formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 5]
```diff
@@ -1081,14 +1082,45 @@ public int compare(Tag o1, Tag o2) {
...
 			String descriptionPath = String.format(descriptionPathPattern, tag.getName());
 			try {
-				descriptionResource = registry.get(descriptionPath);
+				if (govRegistry.resourceExists(descriptionPath)) {
+					descriptionResource = govRegistry.get(descriptionPath);
+				}
 			} catch (RegistryException e) {
-				log.warn(String.format("Cannot get the description for the tag '%s'", tag.getName()));
+				//warn and proceed to the next tag
+				log.warn(String.format("Error while querying the existence of the description for the tag '%s'", tag.getName()));
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: log with exception

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 6]
```diff
@@ -1096,19 +1128,28 @@ public int compare(Tag o1, Tag o2) {
 				try {
 					String description = new String((byte[]) descriptionResource.getContent());
 					tag.setDescription(description);
-				} catch (Exception e) {
-					handleException(String.format("Cannot read content of %s", descriptionPath), e);
+
+				} catch (ClassCastException e) {
+					//added warnings as it can then proceed to load rest of resources/tags
+					log.warn(String.format("Cannot cast content of %s to byte[]", descriptionPath),
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: log with exception unless there is a specific reason

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store/site/themes/fancy/templates/api/api-listing-tag-wise/template.jag [hunk 7]
```diff
@@ -8,8 +8,15 @@
 
         function getThumbnailUrl(tag){
             
+            var urlPrefix = "";
+            var tenantDomain = jagg.getTenantDomain();
+            if (tenantDomain == "carbon.super"){
```
Comment (harsha89) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store/site/themes/fancy/templates/api/api-listing-tag-wise/template.jag]: slight formatting issue

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store/site/themes/default/templates/api/api-listing-tag-wise/template.jag [hunk 8]
```diff
@@ -8,8 +8,15 @@
 
         function getThumbnailUrl(tag){
             
+            var urlPrefix = "";
+            var tenantDomain = jagg.getTenantDomain();
+            if (tenantDomain == "carbon.super"){
```
Comment (harsha89) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store/site/themes/default/templates/api/api-listing-tag-wise/template.jag]: slight formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java (hunk 2 above)
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: This has been fixed later in https://github.com/wso2/carbon-apimgt/commit/e76f1f537c991b6f56c4552bb24e8cff807abf9b

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java (hunk 3 above)
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: This has been fixed later in https://github.com/wso2/carbon-apimgt/commit/0e6afa31d36b155af32340114212c024d5732eb9

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java (hunk 4 above)
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: This has been fixed later in https://github.com/wso2/carbon-apimgt/commit/0e6afa31d36b155af32340114212c024d5732eb9

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store/site/themes/fancy/templates/api/api-listing-tag-wise/template.jag (hunk 7 above)
Comment (malinthaprasan) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store/site/themes/fancy/templates/api/api-listing-tag-wise/template.jag]: related code has been removed by https://github.com/malinthaprasan/carbon-apimgt/commit/0e6afa31d36b155af32340114212c024d5732eb9

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java (hunk 1 above)
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: Fixed in #1494(https://github.com/wso2/carbon-apimgt/commit/b672e7ec2a5973ae34af2999490837d9333cafc4)

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java (hunk 5 above)
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: Fixed in #1494(https://github.com/wso2/carbon-apimgt/commit/b672e7ec2a5973ae34af2999490837d9333cafc4)

**Pull request merged**